cache/in-memory/src/permission.rs

```diff
@@ -426,7 +435,16 @@ impl<'a> InMemoryCachePermissions<'a> {
         let calculator =
             PermissionCalculator::new(guild_id, user_id, everyone, assigned.as_slice());
 
-        Ok(calculator.root())
+        let permissions = calculator.root();
+
+        if !permissions.contains(Permissions::ADMINISTRATOR)
+            && self.is_communication_disabled(user_id, guild_id)
+        {
+            return Ok(permissions
+                .intersection(Permissions::VIEW_CHANNEL | Permissions::READ_MESSAGE_HISTORY));
+        }
```

Comment:
It looks like we have some duplicated logic, could we abstract this out to a method or function?